{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(action='ignore')\n",
    "import os\n",
    "os.chdir('/path/to/your/directory')\n",
    "import sys\n",
    "sys.path.append('./code')\n",
    "import eval_main\n"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Import Data & Preprocess",
   "id": "9880ae91e6eee231"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Input Data\n",
    "1. Pairwise replicates of the same sample, B1-B2, B2-B3, and so on\n",
    "2. spatial coordinates of pairwise alignment/registration samples - mapped coordinates of query sample to ref sample coordinate system\n",
    "3. feature expression matrix, e.g.: gene expression matrix, image feature matrix\n",
    "\n",
    "### Input Preparation\n",
    "1. __Spatial information__: pixel coordinates of the reference section & ref-based(mapped) warped pixel coordinates of the adjacent query section.\n",
    "    - The ref file should have two columns, namely 'x' and 'y', which store the original spatial coordinates, and indexed by spot id.\n",
    "    - The qry file should have two columns, namely 'x' and 'y', which store the warped spatial coordinates, and indexed by spot id.\n",
    "2. __Feature expression matrices__: Reference feature expression matrix & Query feature expression matrix.\n",
    "    - The expression matrix should be a nxm matrix, where it has n spots and m features.\n",
    "    - The default format to read gene expression matrix is `.tsv`; the default format to read any other feature matrix is `.csv`\n"
   ],
   "id": "8ae0365c3e432a18"
  },
  {
   "metadata": {
    "jupyter": {
     "is_executing": true
    }
   },
   "cell_type": "code",
   "source": [
    "# parameter settings\n",
    "ref_id = 'B1'\n",
    "ref_type = 'ref'\n",
    "qry_id = 'B2'\n",
    "qry_type = 'query'\n",
    "total_count = 1e4\n",
    "top_n_svg = 500\n",
    "spatial_dist = 500\n",
    "\n",
    "# directory settings\n",
    "spatial_coords_dir = './sample_data/spatial_coords'\n",
    "feat_dir = './sample_data/features'"
   ],
   "id": "333335c2f8292b24",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "feature_type = 'gene_expression'\n",
    "gene_expression_cost = eval_main.compute_weighted_cost(ref_id, qry_id, spatial_coords_dir, feat_dir,\n",
    "                          feature_type = feature_type,\n",
    "                          top_n_svg = top_n_svg,\n",
    "                          d = spatial_dist,\n",
    "                          total_count = total_count,\n",
    "                          verbose = True)"
   ],
   "id": "7c3338467050ab1e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "feature_type = 'image_features'\n",
    "image_features_cost = eval_main.compute_weighted_cost(ref_id, qry_id, spatial_coords_dir, feat_dir,\n",
    "                          feature_type = feature_type,\n",
    "                          top_n_svg = top_n_svg,\n",
    "                          d = spatial_dist,\n",
    "                          total_count = total_count,\n",
    "                          verbose = True)"
   ],
   "id": "9896b8a2c5dc37ce"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
